contracts/PixelmonEvolution/PixelmonEvolution.sol

```diff
@@ -0,0 +1,350 @@
+    /// @param evolutionStartingId next tokenId to be minted for evolution
+    function setNextEvolvePixelmonId(uint256 evolutionStartingId) external onlyOwner {
+        nextEvolvePixelmonId = evolutionStartingId;
+    }
+
+    /// @notice Sets Serum token contract address
+    /// @dev This function can only be executed by the contract owner
+    /// @param serumAddress Serum token Contract
+    function setSerumAddress(address serumAddress) external onlyOwner {
+        SERUM_CONTRACT = IERC1155(serumAddress);
+    }
+
+    /// @notice Sets Signer wallet address
+    /// @dev This function can only be executed by the contract owner
+    /// @param signer Signer wallet address for signature verifition
+    function setSignerAddress(address signer) external onlyOwner {
+        SIGNER = signer;
+    }
+
+    /// @notice Returns the hash integer value from an array of integer
+    /// @dev This is a view function and have no gas cost
+    /// @param array array of integer
+    function getHashIntFromArray(uint256[] memory array) public pure returns (uint256) {
+        return uint256(keccak256(abi.encodePacked(array)));
+    }
+
+    /// @notice Burns Serum token and mint evolved Pixelmon token
+    /// @dev Owner of Pixelmon token and Serum token can call with proper signature
+    /// @param pixelmonTokenIds List of Pixelmon tokenId
+    /// @param serumIds List of Serum token Id
+    /// @param serumAmounts List of Serum token amount
+    /// @param evolutionStage Evolution stage for minting evolve Pixelmon
+    /// @param nonce Nonce for the transaction to evolve pixelmon
+    /// @param stakedFor Staking time period in second
+    /// @param signature Signature from signer wallet
+    function evolvePixelmon(
+        uint256[] memory pixelmonTokenIds,
+        uint256[] memory serumIds,
+        uint256[] memory serumAmounts,
+        uint256 evolutionStage,
+        uint256 nonce,
+        uint256 stakedFor,
+        bytes calldata signature
+    ) external nonReentrant noContracts {
+        if (serumIds.length != serumAmounts.length) {
+            revert InvalidData();
+        }
+
+        uint256 totalEvolveAmount;
+        for (uint256 index = 0; index < serumAmounts.length; index++) {
+            totalEvolveAmount += serumAmounts[index];
+        }
+
+        if (totalEvolveAmount != pixelmonTokenIds.length) {
+            revert InvalidAmount();
+        }
+
+        if (nonce != nonces[msg.sender]) {
+            revert InvalidNonce();
+        }
+        unchecked {
+            nonces[msg.sender]++;
+        }
+
+        address signer = recoverSignerFromSignature(
+            pixelmonTokenIds,
+            serumIds,
+            serumAmounts,
+            evolutionStage,
+            nonce,
+            stakedFor,
+            msg.sender,
+            signature
+        );
+
+        if (signer != SIGNER) {
+            revert InvalidSignature();
+        }
```

Comment:
Is there any issue if we change signer address while some Pixelmon already evolved and staked in the contract?
Will user still be able to evolve and claim the Pixelmon?

Reply:
No. Signer is only for checking the signature during the staking.

If we change the signer after staking, it will not have any effect on the claim. And the new staking will verify with the new signer.